File: src/main/java/me/ag2s/epublib/domain/Date.java
package me.ag2s.epublib.domain;

import java.io.Serializable;
import java.text.SimpleDateFormat;
import java.util.Locale;

public class Date implements Serializable {
   private static final long serialVersionUID = 7533866830395120136L;
   private Event event;
   private String dateString;

   public Date() {
      this(new java.util.Date(), Date.Event.CREATION);
   }

   public Date(java.util.Date date) {
      this(date, (Event)null);
   }

   public Date(String dateString) {
      this(dateString, (Event)null);
   }

   public Date(java.util.Date date, Event event) {
      this((new SimpleDateFormat("yyyy-MM-dd", Locale.US)).format(date), event);
   }

   public Date(String dateString, Event event) {
      this.dateString = dateString;
      this.event = event;
   }

   public Date(java.util.Date date, String event) {
      this((new SimpleDateFormat("yyyy-MM-dd", Locale.US)).format(date), event);
   }

   public Date(String dateString, String event) {
      this(checkDate(dateString), Date.Event.fromValue(event));
      this.dateString = dateString;
   }

   private static String checkDate(String dateString) {
      if (dateString == null) {
         throw new IllegalArgumentException("Cannot create a date from a blank string");
      } else {
         return dateString;
      }
   }

   public String getValue() {
      return this.dateString;
   }

   public Event getEvent() {
      return this.event;
   }

   public void setEvent(Event event) {
      this.event = event;
   }

   public String toString() {
      return this.event == null ? this.dateString : "" + this.event + ":" + this.dateString;
   }

   public static enum Event {
      PUBLICATION("publication"),
      MODIFICATION("modification"),
      CREATION("creation");

      private final String value;

      private Event(String v) {
         this.value = v;
      }

      public static Event fromValue(String v) {
         for(Event c : values()) {
            if (c.value.equals(v)) {
               return c;
            }
         }

         return null;
      }

      public String toString() {
         return this.value;
      }
   }
}

File: src/main/kotlin/me/ag2s/epublib/domain/Date.kt
package me.ag2s.epublib.domain

import java.io.Serializable
import java.text.SimpleDateFormat
import java.util.Locale

class Date : Serializable {
    private var event: Event?
    private var dateString: String?
    constructor() : this(java.util.Date(), Event.CREATION)
    constructor(date: java.util.Date) : this(date, null as Event?)
    constructor(dateString: String?) : this(dateString, null as Event?)
    constructor(date: java.util.Date, event: Event?) : this(SimpleDateFormat("yyyy-MM-dd", Locale.US).format(date), event)
    constructor(dateString: String?, event: Event?) { this.dateString = dateString; this.event = event }
    constructor(date: java.util.Date, event: String?) : this(SimpleDateFormat("yyyy-MM-dd", Locale.US).format(date), event)
    constructor(dateString: String?, event: String?) : this(checkDate(dateString), Event.fromValue(event)) { this.dateString = dateString }
    fun getValue(): String? = dateString
    fun getEvent(): Event? = event
    fun setEvent(event: Event?) { this.event = event }
    override fun toString(): String = if (event == null) dateString as String else "$event:$dateString"
    enum class Event(private val value: String) { PUBLICATION("publication"), MODIFICATION("modification"), CREATION("creation"); override fun toString() = value; companion object { @JvmStatic fun fromValue(value: String?): Event? = entries.firstOrNull { it.value == value } } }
    companion object { private const val serialVersionUID = 7533866830395120136L; private fun checkDate(dateString: String?): String { require(dateString != null) { "Cannot create a date from a blank string" }; return dateString } }
}
